src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs

```diff
@@ -0,0 +1,223 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Azure.WebJobs.Host.Bindings.Path;
+using System.Collections.Generic;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Host.Bindings
+{
+    // Clone an attribute and resolve it. 
+    // This can be tricky since some read-only properties are set via the constructor. 
+    // This assumes that the property name matches the constructor argument name. 
+    internal class AttributeCloner<TAttribute>
+        where TAttribute : Attribute
+    {
+        private readonly TAttribute _source;
+        private readonly ConstructorInfo _bestCtor;
+        private readonly Func<IReadOnlyDictionary<string, object>, object>[] _bestCtorArgBuilder;
+        private readonly INameResolver _nameResolver;
+
+        // If no name resolver is specified, then any %% becomes an error. 
+        class EmptyNameResolver : INameResolver
+        {
+            public string Resolve(string name)
+            {
+                return null;
+            }
+        }
+
+        public AttributeCloner(TAttribute source, INameResolver nameResolver = null)
+        {
+            _nameResolver = nameResolver ?? new EmptyNameResolver();
+            _source = source;
+
+            Type t = typeof(TAttribute);
+
+            Dictionary<string, PropertyInfo> availableParams = new Dictionary<string, PropertyInfo>(StringComparer.OrdinalIgnoreCase);
+            foreach (var prop in t.GetProperties(BindingFlags.Instance | BindingFlags.Public))
+            {
+                var objValue = prop.GetValue(_source);
+                if (objValue != null)
+                {
+                    availableParams[prop.Name] = prop;
+                }
+            }
+
+            int longestMatch = -1;
+
+            // Pick the ctor with the longest parameter list where all parameters are matched. 
+            var ctors = t.GetConstructors(BindingFlags.Public | BindingFlags.Instance);
+            foreach (var ctor in ctors)
+            {
+                var ps = ctor.GetParameters();
+                int len = ps.Length;
+
+                var getArgFuncs = new Func<IReadOnlyDictionary<string, object>, object>[len];
+
+                bool ok = true;
```

Comment:
nit: please rename `ok` to something more descriptive so I know what it is for :)